### ISSUE TITLE
websocket eof exception

### ISSUE THREAD

**Issue by KryptomineCH** [dkackman/chia-dotnet#62]
**Describe the bug**
When trying to connect to websocket, the connection fails with the following exception:
``` exception
Class Initialization method chia.dotnet.tests.DaemonTests.Initialize threw exception. 
System.Net.Http.HttpRequestException: 
    System.Net.Http.HttpRequestException: 
        The SSL connection could not be established, see inner exception. 
        ---> System.IO.IOException:  Received an unexpected EOF or 0 bytes from the transport stream..
```

With Testing with the cloned github:
this is the connectionstring:
``` c#
await _webSocket.ConnectAsync(new Uri("wss://localhost:55400/"), cancellationToken).ConfigureAwait(false);
```

Minimal reproducible example (with chia-dotnet nuget package)
1. create unit test project
2. add nuget package chia-dotnet
3. create the following unit test:
``` c#
[Fact]
public void TestChiaDotNet()
{
    var endpoint = Config.Open().GetEndpoint("daemon");
    using var rpcClient = new WebSocketRpcClient(endpoint);
    rpcClient.Connect().Wait();

    var daemon = new DaemonProxy(rpcClient, "unit_tests");
    daemon.RegisterService().Wait();

    var fullNode = new FullNodeProxy(rpcClient, "unit_tests");
    var state = fullNode.GetBlockchainState().Result;
}
```

**Expected behavior**
Connection gets established

-> do I need to gonfigure something so the demon is listening?

**Desktop (please complete the following information):**
 - OS: Win 10
 - latest github repo

**Comment by dkackman** [comment]
Based on the error message it might be the cert files.

The daemon needs to be running of course, and the port needs to be accessible.

Is this unit test code running on the same machine as the daemon?

**Comment by KryptomineCH** [comment]
yes, both running on the same machine. Interestingly, I got the same error with your library and with my own code.
I also came to the conclusion that its something with the certificates. 

though I managed to connect to other endpoints with a standard http request. (needed to do a lot of digging until I found the correct certificate import code)

**Comment by dkackman** [comment]
did you get it working?

**Comment by KryptomineCH** [comment]
not with websocket. only with single http requests
```
internal static X509Certificate2 GetCertificate(Endpoint endpoint)
{
    X509Certificate2 certificate = null;
    if (Certificates.TryGetValue(endpoint, out certificate))
    {
        return certificate;
    }
    DirectoryInfo endpointCertificateDirectory = new DirectoryInfo(Path.Combine(CertificatesFolder, endpoint.ToString()));
    FileInfo[] files = endpointCertificateDirectory.GetFiles();
    FileInfo certFile = null;
    FileInfo keyFile = null;
    foreach(FileInfo file in files)
    {
        if (!file.Name.StartsWith("private")) continue;
        if (file.Name.EndsWith(".crt")) certFile = file;
        else if (file.Name.EndsWith(".key")) keyFile = file;
    }
    if (certFile == null || keyFile == null)
    {
        throw new Exception($"certificate file for endpoint {endpoint} not found!");
    }
    certificate = X509Certificate2.CreateFromPemFile(certFile.FullName, keyFile.FullName);
    certificate = new X509Certificate2(certificate.Export(X509ContentType.Pfx));
    Certificates[endpoint] = certificate;
    return certificate;
}
```

**Comment by dkackman** [comment]
hmm... there is a [similar bit of code in this library](https://github.com/dkackman/chia-dotnet/blob/main/src/chia-dotnet/CertLoader.cs), which should work for both http ans ws. You could try building up the `EndpointInfo` object manually rather than relying on what's in the config. Or make sure that what you get back from ` var endpoint = Config.Open().GetEndpoint("daemon");` points to the correct cert locations.

**Comment by KryptomineCH** [comment]
I have tried your cert loader. didnt work for me.
also made sure the endpoints were correct (you hardcoded your username instead of Environment.GetSpecialfolder at least in the unit tests)

 -> are you using linux?
 Apparently .net linux uses openssl (which has no issues with chia client)
while windows.net uses windows Schannel (which is a pain in the *)

**Comment by dkackman** [comment]
> I have tried your cert loader. didnt work for me. also made sure the endpoints were correct (you hardcoded your username instead of Environment.GetSpecialfolder at least in the unit tests)
> 
yeesh. yeah i can see myself doing that in the unit tests. laziness. im confident i didn't do it elsewhere though.

I'm not sure what else to suggest. You could package up a minimal example of how you have things coded and setup and i can see if i notice anything.

**Comment by KryptomineCH** [comment]
litterally with your code, no changes except the location of the key files.

But I think im going to close this. Im having great success with straight htt rest library.
As it's localhost, there isnt much performance impact compared to websocket.
If you still want to check the library out (so far key-management is complete, others are integrated but no integration tests so far):
https://github.com/KryptomineCH/Chia-Client-API 
dependency: https://github.com/KryptomineCH/CHIA-RPC

**Comment by dkackman** [comment]
this library support http as well BTW. https://github.com/dkackman/chia-dotnet/blob/main/src/chia-dotnet/HttpRpcClient.cs

If you're interested in combining efforts I'd love to figure out why this isn't working for you and collaborating on this code. I haven't had much time to work on this of late.

**Comment by KryptomineCH** [comment]
I designed it such that the service interfaces are reused between websocket and http so you can switch between the two with minimal effort.
-> why would you do that if one is working?

> I haven't had much time to work on this of late.
I have seen that. I thought that was the original reason for your python implementation which supports 1.6.1 and you kept this one up for legacy use.

- I have searched plenty in the internet specifically about tls 1.3 only to find that windows doesnt properly support tls 1.3
Linux implementation uses openssl which does not have any issues.
- I think it might have worked on a previous release before the security close for Poodle attack

If you dont mind, it would be appreciated if you can run one of your unit tests against v 1.6.1 on windows.

What I found was that daemon only has private keys in its folder but I dont know if that would have any effect. the other endpoints seem to have public & private keys. On 2 machines the same result 
![image](https://user-images.githubusercontent.com/117320700/207434722-bd7fd225-9c3e-4878-93ac-ea86ac204c40.png)

**Comment by dkackman** [comment]
> why would you do that if one is working?

To provide flexibility for the end user and maintain the portability of code between uses and use cases. 

> If you dont mind, it would be appreciated if you can run one of your unit tests against v 1.6.1 on windows.
Will do. All of my servers are linux and I've tested this with both linux and windows clients, but never with a windows server.

**Comment by greimela** [comment]
I ran into the same issue, when trying to connect to the Chia daemon websocket.
Exactly the same error message: `Received an unexpected EOF or 0 bytes from the transport stream.`

I already tried all the things related to ClientCertificates and RemoteCertificateValidation, but it simply doesn't want to work...
My python example is able to connect to the same websocket with the same certificates without issue.

I'm using .NET core on MacOS.
Any ideas what I could try to fix this?

**Comment by greimela** [comment]
Update: found the issue, but no resolution yet. The problem is that the demon forces TLS 1.3, which Dotnet is not able to handle for me...

**Comment by KryptomineCH** [comment]
Hello Greimela,

as a simple option, I would recommend to use the rest api instead. 
Yes, websocket is faster in theory but it is likely in the local network anyways and the blockchain is not so fast that the speed would be critcal in any regard I think.

As far as I understood it, It is likely due to tls 1.3 not beeing available on windows 10 (you are running windows 10 I guess)
Windows 11 would likely resolve the issue for you

**Comment by dkackman** [comment]
reopening as I never did get to testing this on windows 10 and the stale bot closed it.

**Comment by dkackman** [comment]
> Update: found the issue, but no resolution yet. The problem is that the demon forces TLS 1.3, which Dotnet is not able to handle for me...

Can you try to enable tls 1.3 system wide? 
https://www.itechtics.com/tls-1-3/

This will be my first step to try to find a workaround but I'll I need to get a VM going so might be worth your while to try it too.

**Comment by greimela** [comment]
I'm on MacOS, it looks like this is the issue: https://github.com/dotnet/runtime/issues/1979
It simply doesn't support TLS 1.3... 
Sorry for the noise, doesn't look like you can do anything here.

@KryptomineCH 
Yes, looks like I will poll via RPC for now.

**Comment by dkackman** [comment]
> I'm on MacOS, it looks like this is the issue: [dotnet/runtime#1979](https://github.com/dotnet/runtime/issues/1979) It simply doesn't support TLS 1.3... Sorry for the noise, doesn't look like you can do anything here.

No problem. 

Now I am suddenly curious as to why chia's wss endpoint requires 1.3 and their https does not. I would expect the version support to be the same one way or the other, so perhaps it's a misconfiguration bug of wss on their side that could be fixed. I'll poke around in their code when I get to this over the weekend.

**Comment by KryptomineCH** [comment]
I was in contact with them already but got the note that this is intentional.
I believe I tried that before but it did not work... https://www.itechtics.com/tls-1-3/

**Comment by dkackman** [comment]
@KryptomineCH @greimela do either of you see this message in your logs after you try it?

`Deprecation Warning: Your version of SSL (%s) does not support TLS1.3.
                    A future version of Chia will require TLS1.3.`

I might do a PR to chia to allow 1.2 via config setting. It may be that or trying to figure out why it isn't falling back or why 1.2 isn't working when it does.

**Comment by dkackman** [comment]
Finally got a windows 10 environment running and `ssl.HAS_TLSv1_3` is indeed indicating that tls 1.3 is available. When I forcibly set `self.ssl_context.minimum_version = ssl.TLSVersion.TLSv1_2` chia.dotnet works as expected.

Not sure if its the same case on macOS but, short of figuring out what's up with `ssl.HAS_TLSv1_3` the only fix I can think of is a config flag to use 1.2 by default in the daemon. I'll do a PR and see what they think. 

@greimela if you comment out this block of code in `chia\daemon\server.py` line 182 can you see if that works on macOs?

```
if ssl.HAS_TLSv1_3:
    try:
        self.ssl_context.minimum_version = ssl.TLSVersion.TLSv1_3
    except ValueError:
        # in case the attempt above confused the config, set it again (likely not needed but doesn't hurt)
        self.ssl_context.minimum_version = ssl.TLSVersion.TLSv1_2
```

**Comment by greimela** [comment]
Yes, I tried that, it works with TLS 1.2!

Would love to have a config option for that.

**Comment by dkackman** [comment]
> Yes, I tried that, it works with TLS 1.2!
> 
> Would love to have a config option for that.

Take a look at https://github.com/Chia-Network/chia-blockchain/pull/16731. Pretty straightforward.

Also, fortuitously, while testing this, I found that chia now serializes new connection signing keys as PKCS 8 instead of RSA, Not sure exactly when that changed but was able to add support for that along the way.

**Comment by KryptomineCH** [comment]
I will test on windows this weekend.
added the config flag to my node already.

**Comment by KryptomineCH** [comment]
Edit: I noticed it's in 2.1.2-rc1

so probably I have to wait a couple of days still. 

Dang.

**Comment by dkackman** [comment]
fixed in chia 2.1.12
https://github.com/Chia-Network/chia-blockchain/releases/tag/2.1.2
